src/trace_writer.ml

```diff
@@ -821,121 +845,142 @@ let write_event (T t) event =
       | Some ip -> is_kernel_address ip
     in
     end_of_thread t thread_info ~time ~is_kernel_address
-  | Ok (Power { thread = _; time = _; freq }) ->
-    write_counter
-      t
-      ~thread
-      ~name:"CPU"
-      ~time
-      ~args:Tracing.Trace.Arg.[ "freq (MHz)", Int freq ]
-  | Ok (Trace event) ->
-    let { Event.Ok.Trace.thread = _ (* Already used this to look up thread info. *)
-        ; time = _ (* Already in scope. Also, this time hasn't been [map_time]'d. *)
-        ; kind
-        ; trace_state_change
-        ; src
-        ; dst
-        }
-      =
-      event
-    in
-    Ocaml_hacks.track_executed_pushtraps_and_poptraps_in_range
-      t
-      thread_info
-      ~src
-      ~dst
-      ~time;
-    (match kind, trace_state_change with
-    | Some Call, (None | Some End) -> call t thread_info ~time ~location:dst
-    | ( Some (Call | Syscall | Return | Hardware_interrupt | Iret | Sysret | Jump)
-      , Some Start )
-    | Some (Hardware_interrupt | Jump), Some End ->
-      raise_s
-        [%message
-          "BUG: magic-trace devs thought this event was impossible, but you just proved \
-           them wrong. Please report this to \
-           https://github.com/janestreet/magic-trace/issues/"
-            (event : Event.Ok.Trace.t)]
-    | None, Some End -> call t thread_info ~time ~location:Event.Location.untraced
-    | Some Syscall, Some End ->
-      (* We should only be getting these under /u *)
-      assert_trace_mode t outer_event [ Userspace ];
-      call t thread_info ~time ~location:Event.Location.syscall
-    | Some Return, Some End -> call t thread_info ~time ~location:Event.Location.returned
-    | Some Return, None ->
-      Ocaml_hacks.ret_track_exn_data t thread_info ~time;
-      check_current_symbol t thread_info ~time dst
-    | None, Some Start ->
-      (* Might get this under /u, /k, and /uk, but we need to handle them all
+  | Ok
+      { Event.Ok.thread = _ (* Already used this to look up thread info. *)
+      ; time = _ (* Already in scope. Also, this time hasn't been [map_time]'d. *)
+      ; data
+      } ->
+    (match data with
```

Comment:
Nit: instead of having an inner match, perhaps this code would look a bit better (less indented) if the outer block matched on

```ocaml
| Ok
      { Event.Ok.thread = _ (* Already used this to look up thread info. *)
      ; time = _ (* Already in scope. Also, this time hasn't been [map_time]'d. *)
      ; data = Power { freq }
      } ->
```

etc.

Reply:
Updated